Improve DateHeaders in MockServletRequest/Response

bclozel opened the pull request:
Prior to this change, calling the `setDateHeader` method on a
Spring Test MockHttpServletResponse instance would just store the given
long value in a Map, not writing it as a formatted date String.
Also, calling `getDateHeader` on a MockHttpServletRequest would not
support date strings and could not parse those values.

This can be problematic when testing features related to date headers
such as "Expires", "If-Modified-Since", "Last-Modified", etc.

This commit adds formatting and parsing capabilities to Servlet Mocks
for date strings in HTTP headers.

When formatting dates to Strings, the date format used is the one
preferred by the HTTP RFC. When parsing date Strings, multiple date
formats are supported for better compatibility.

Issue: SPR-11912

bclozel committed:
Improve DateHeaders in MockServletRequest/Response

Prior to this change, calling the `setDateHeader` method on a
Spring Test MockHttpServletResponse instance would just store the given
long value in a Map, not writing it as a formatted date String.
Also, calling `getDateHeader` on a MockHttpServletRequest would not
support date strings and could not parse those values.

This can be problematic when testing features related to date headers
such as "Expires", "If-Modified-Since", "Last-Modified", etc.

This commit adds formatting and parsing capabilities to Servlet Mocks
for date strings in HTTP headers.

When formatting dates to Strings, the date format used is the one
preferred by the HTTP RFC. When parsing date Strings, multiple date
formats are supported for better compatibility.

Issue: SPR-11912

sbrannen commented on spring-test/src/main/java/org/springframework/mock/web/MockHttpServletRequest.java:
```diff
@@ -912,6 +928,9 @@ public long getDateHeader(String name) {
+		}
 		else if (value != null) {
 			throw new IllegalArgumentException(
 					"Value for header '" + name + "' is neither a Date nor a Number: " + value);
```
Need to update exception message as follows:

`"Value for header '" + name + "' is not a Date, Number, or String: " + value`

sbrannen commented on spring-test/src/main/java/org/springframework/mock/web/MockHttpServletRequest.java:
```diff
@@ -921,6 +940,21 @@ else if (value != null) {
+				// ignore
+			}
+		}
+		throw new IllegalArgumentException("Cannot parse date value \"" + value +
```
AFAIK, in Spring we typically enclose _quoted_ values within exception messages in single quotes.

sbrannen commented:
What do you think about adding a `setDateHeader()` method in `MockHttpServletRequest` as suggested in the JIRA issue?

sbrannen commented on spring-web/src/test/java/org/springframework/mock/web/test/MockHttpServletRequest.java:
```diff
@@ -912,6 +927,9 @@ public long getDateHeader(String name) {
+		}
 		else if (value != null) {
 			throw new IllegalArgumentException(
 					"Value for header '" + name + "' is neither a Date nor a Number: " + value);
```
same as previous comment from `spring-test` mock.

sbrannen commented on spring-web/src/test/java/org/springframework/mock/web/test/MockHttpServletRequest.java:
```diff
@@ -921,6 +939,21 @@ else if (value != null) {
+				// ignore
+			}
+		}
+		throw new IllegalArgumentException("Cannot parse date value \"" + value +
```
same as previous comment from `spring-test` mock.

bclozel commented:
Thanks @sbrannen for this review! I've fixed all the issues you pointed at.
I've updated the javadoc for `addHeader` with the supported types (the implementation details it was describing were wrong anyway). The supported date formats are listed on the `getDateHeader` method.

sbrannen commented:
Good deal. Thanks!
